otter/convergence/steps.py

```diff
@@ -161,6 +163,55 @@ def report_failure(result):
                                    error=report_failure)
 
 
+class UnexpectedServerStatus(Exception):
+    """
+    An exception to be raised when a server is found in an unexpected state.
+    """
+    def __init__(self, server_id, status, expected_status):
+        super(UnexpectedServerStatus, self).__init__(
+            'Expected {server_id} to have {expected_status}, '
+            'has {status}'.format(server_id=server_id,
+                                  status=status,
+                                  expected_status=expected_status)
+        )
+        self.server_id = server_id
+        self.status = status
+        self.expected_status = expected_status
+
+
+def delete_and_verify(server_id):
+    """
+    Check the status of the server to see if it's actually been deleted.
+    Succeeds only if it has been either deleted (404) or acknowledged by Nova
+    to be deleted (task_state = "deleted").
+
+    Note that ``task_state`` is in the server details key
+    ``OS-EXT-STS:task_state``, which is supported by Openstack but available
+    only when looking at the extended status of a server.
+    """
+
+    def check_task_state((resp, json_blob)):
+        server_details = json_blob['server']
+        is_deleting = server_details.get("OS-EXT-STS:task_state", "")
+        if is_deleting.strip().lower() != "deleting":
+            raise UnexpectedServerStatus(server_id, is_deleting, "deleting")
+
+    def verify((_type, error, traceback)):
+        if _type is not APIError or error.code != 204:
+            raise error
+        ver_eff = service_request(
+            ServiceType.CLOUD_SERVERS, 'GET',
+            append_segments('servers', server_id, 'details'),
+            success_pred=has_code(200))
+        return ver_eff.on(check_task_state).on(
+            error=catch(APIError, lambda (t, e, _): e.code == 404))
+
+    return service_request(
+        ServiceType.CLOUD_SERVERS, 'DELETE',
+        append_segments('servers', server_id),
+        success_pred=has_code(404)).on(error=verify)
```

Comment:
You can use .on(error=catch(APIError, verify)) here to guard against non-APIError. You can update requirements.txt to effect==0.1a14 to get this function (don't worry about conflicts with other PRs).